pkg/admission/crdnooverlappinggvr/crdnooverlappinggvr_admission.go

```diff
@@ -0,0 +1,131 @@
+package crdnooverlappinggvr
+
+import (
+	"context"
+	"fmt"
+	kcpinformers "github.com/kcp-dev/kcp/pkg/client/informers/externalversions"
+	"io"
+
+	apisv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/apis/v1alpha1"
+	apislisters "github.com/kcp-dev/kcp/pkg/client/listers/apis/v1alpha1"
+	"github.com/kcp-dev/logicalcluster"
+
+	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apiserver/pkg/admission"
+	"k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/client-go/tools/clusters"
+)
+
+const (
+	PluginName = "apis.kcp.dev/CRDNoOverlappingGVR"
+)
+
+func Register(plugins *admission.Plugins) {
+	plugins.Register(PluginName,
+		func(_ io.Reader) (admission.Interface, error) {
+			return &crdNoOverlappingGVRAdmission{
+				Handler: admission.NewHandler(admission.Create),
+			}, nil
+		})
+}
+
+type crdNoOverlappingGVRAdmission struct {
+	*admission.Handler
+
+	apiBindingLister        apislisters.APIBindingLister
+	apiExportLister         apislisters.APIExportLister
+	apiResourceSchemaLister apislisters.APIResourceSchemaLister
+}
+
+// Ensure that the required admission interfaces are implemented.
+var _ = admission.ValidationInterface(&crdNoOverlappingGVRAdmission{})
+
+func (p *crdNoOverlappingGVRAdmission) SetKcpInformers(informers kcpinformers.SharedInformerFactory) {
+	p.SetReadyFunc(func() bool {
+		return informers.Apis().V1alpha1().APIBindings().Informer().HasSynced() &&
+			informers.Apis().V1alpha1().APIExports().Informer().HasSynced() &&
+			informers.Apis().V1alpha1().APIResourceSchemas().Informer().HasSynced()
+	})
+	p.apiBindingLister = informers.Apis().V1alpha1().APIBindings().Lister()
+	p.apiExportLister = informers.Apis().V1alpha1().APIExports().Lister()
+	p.apiResourceSchemaLister = informers.Apis().V1alpha1().APIResourceSchemas().Lister()
+}
+
+// Validate checks if the given CRD's Group and Resource don't overlap with other CRDs exposed via APIBindings
+func (p *crdNoOverlappingGVRAdmission) Validate(ctx context.Context, a admission.Attributes, _ admission.ObjectInterfaces) error {
+	if a.GetResource().GroupResource() != apiextensions.Resource("customresourcedefinitions") {
```

Comment:
I think I need to find a way to skip `CRDs` created by the APIBinding controller

Reply:
or actually, we should make a distinction between CRDs created for the binding and the ones created by an admin

Reply:
the ones from APIBinding live in the `system:bounds-crds` logical cluster.

Reply:
there seem to be no way of identifying the target cluster from a CRD in the logical cluster

Reply:
I think the APIBinding CRDs in `system:bound-crds` you can ignore here as long as you have the GVRs for them from APIBinding.status.

The `system:admin` CRDs matter too here (today that's all from https://github.com/kcp-dev/kcp/tree/main/config/crds, but most of them will turn into APIBindings soon).